### ISSUE TITLE
dart2js: Inline single-reference constants into other constants

### ISSUE THREAD

**Issue by rakudrama** [dart-lang/sdk#25230]
Some constants in the constant pool contain references to other constants where the reference is the only reference to that constant.
The constant expression could be moved to the site of the reference and avoid needing the name.
Example: this is often true of the list of keys for a map constant.

We might have to take care and keep some references to limit the size of any one expression.

**Comment by rakudrama** [comment]
With deferred loaded part files, there are 

1. Constants referred to by the constant pool of another part file
2. Constants referred to by code in another part file
3. Constants referred to only by code (and possibly constants) in the current part file
4. Constants referred to only by other constants in the current part file
5. Constants referred to only once by other constants in the current part file.

All these are currently handled via globally addressable named constants.

1 and 2 need a name addressable from the other part file (e.g. C.xyz).
3 needs a local variable in scope of the part-local method.
4 needs a local variable internal to the constant pool construction.
5 can be generated in-place within the containing constant.

3 is difficult to achieve in the current framework since we can't predict where constants will end up (inlining and constant-folding move and create references). We could create a deferred JS AST for the references and replace that with a local name if, after generating all code, all references were from the same part, otherwise use the global naming scheme.
